server/src/main/java/org/apache/druid/catalog/model/table/ResolvedExternalTable.java

```diff
@@ -0,0 +1,141 @@
+  protected final Map<String, Object> inputFormatMap;
+  private InputSourceDefn inputSourceDefn;
+  private InputFormatDefn inputFormatDefn;
+
+  /**
+   * Construct a resolved external table by extracting the input source
+   * and input format properties, and converting each to a Java map.
+   * Validates that the input source is present: the format is optional.
+   * <p>
+   * Note: does <i>not</i> resolve the input source and input format
+   * definitions: that is done as a separate step when needed.
+   *
+   * @see {@link #resolve(TableDefnRegistry)}.
+   */
+  public ResolvedExternalTable(final ResolvedTable table)
+  {
+    this.table = table;
+    Map<String, Object> map = table.mapProperty(ExternalTableDefn.SOURCE_PROPERTY);
+    if (map == null || map.isEmpty()) {
+      throw new IAE("%s property is required", ExternalTableDefn.SOURCE_PROPERTY);
+    }
+    this.inputSourceMap = new HashMap<>(map);
+    map = table.mapProperty(ExternalTableDefn.FORMAT_PROPERTY);
+    this.inputFormatMap = map == null ? null : new HashMap<>(map);
+  }
+
+  public ResolvedTable resolvedTable()
+  {
+    return table;
+  }
+
+  /**
+   * Look up the input source type and input format type to find the corresponding
+   * definitions in the table registry. Throws an exception if the types are not
+   * defined. The input source is required, the format is optional.
+   * <p>
+   * Note, for resolution to work, the name of each definition must be the same as
+   * that used as the type key in the serialized JSON.
+   */
+  public ResolvedExternalTable resolve(TableDefnRegistry registry)
+  {
+    String inputSourceType = CatalogUtils.getString(inputSourceMap, InputSource.TYPE_PROPERTY);
+    if (inputSourceType == null) {
+      throw new IAE("Input source type %s is required", InputSource.TYPE_PROPERTY);
+    }
+    inputSourceDefn = registry.inputSourceDefnFor(inputSourceType);
+    if (inputSourceDefn == null) {
+      throw new IAE("Input source type %s is not registered", inputSourceType);
+    }
+    if (inputFormatMap != null) {
+      String inputFormatType = CatalogUtils.getString(inputFormatMap, InputFormat.TYPE_PROPERTY);
+      if (inputFormatType != null) {
+        inputFormatDefn = registry.inputFormatDefnFor(inputFormatType);
+        if (inputFormatDefn == null) {
+          throw new IAE("Input format type %s is not registered", inputFormatType);
+        }
+      }
+    }
+    return this;
+  }
+
+  /**
+   * Validate that the table spec is correct by resolving the definitions, then
+   * converting the JSON to the desired object type. Note that this path requires
+   * special handling: the table spec may be partial, which means it is missing information
+   * needed to create a complete input source. The input source definition defines which
+   * values can be omitted, and defined later in SQL via function parameters. If those
+   * values are missing, then the input source defn should provide dummy values so that
```

Comment:
question: Where are these dummy values that I've seen set in places removed from the definition of the input source, so that we dont potentially run an ingestion with dummy values set?